narwhal/primary/src/primary.rs

```diff
@@ -550,6 +541,221 @@ impl PrimaryReceiverHandler {
         }
         Ok(None)
     }
+
+    async fn process_request_vote(
+        &self,
+        request: anemo::Request<RequestVoteRequest>,
+    ) -> DagResult<RequestVoteResponse> {
+        let network = request
+            .extensions()
+            .get::<anemo::NetworkRef>()
+            .and_then(anemo::NetworkRef::upgrade)
+            .ok_or_else(|| {
+                DagError::NetworkError("Unable to access network to send child RPCs".to_owned())
+            })?;
+
+        let header = &request.body().header;
+        let committee = self.committee.load();
+        header.verify(&committee, self.worker_cache.clone())?;
+
+        // Vote request must come from the Header's author.
+        // DO NOT MERGE: Will enforcing RPC requester is same as Header author cause any issues?
+        let peer_id = request
+            .peer_id()
+            .ok_or_else(|| DagError::NetworkError("Unable to access remote peer ID".to_owned()))?;
+        let peer_network_key = NetworkPublicKey::from_bytes(&peer_id.0).map_err(|e| {
+            DagError::NetworkError(format!(
+                "Unable to interpret remote peer ID {peer_id:?} as a NetworkPublicKey: {e:?}"
+            ))
+        })?;
+        let (peer_authority, _) = committee
+            .authority_by_network_key(&peer_network_key)
+            .ok_or_else(|| {
+                DagError::NetworkError(format!(
+                    "Unable to find authority with network key {peer_network_key:?}"
+                ))
+            })?;
+        ensure!(
+            header.author == *peer_authority,
+            DagError::NetworkError(format!(
+                "Header author {:?} must match requesting peer {peer_authority:?}",
+                header.author
+            ))
+        );
+
+        debug!(
+            "Processing vote request for {:?} round:{:?}",
+            header, header.round
+        );
+
+        // Clone the round updates channel so we can get update notifications specific to
```

Comment:
Looking again on the code it seems that this is the check we had before:

```
        ensure!(
            self.gc_round < header.round,
            DagError::TooOld(header.digest().into(), header.round, self.gc_round)
        );
```

where the `gc_round = consensus_round - gc_depth`

Since this PR is changing quite a lot , I would suggest keeping the checks as the ones we had before and refine them in a follow up PR

Reply:
That check is still done in several places in the RequestVote handler, we just don't have the concept of gc_round there (and as George pointed out it should use narwhal round and not consensus round as it did previously)